CSRF Protection

DarkMetalMouse opened the pull request:
Added CSRF protection with minimal changes to the code.

The protection has two components:
1. Blocking non `application/json` requests to the API. See [Disallowing non-simple requests](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#disallowing-non-simple-requests)
2. For file upload (`<form>`), add a custom header. See [Employing Custom Request Headers for AJAX/API](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#employing-custom-request-headers-for-ajaxapi)

Note that I didn't add CSRF protection to the dashboard. The frontend source code needs to be modified to include the custom header, and I do not have access to the dashboard code.

I have tested uploading files and accessing the API for both the admin and regular APIs

johnmeshulam commented:
Thanks. Regarding the custom header -> I see you added it to one request but there are more than one file upload routes. After this PR is merged do I need to manually add it to each upload?

DarkMetalMouse commented:
I added the header to `UploadFileButton`, it should be present is all file uploads since they all use `UploadFileButton`. Regarding the backend, the `csrfValidator` is called before all requests to anything under `/api/` according to my testing:
![Screenshot 2024-03-25 at 16 41 18](https://github.com/FIRSTIsrael/lems/assets/51059131/f015966d-a017-4222-b212-6b7622f15c79)
![Screenshot 2024-03-25 at 16 41 28](https://github.com/FIRSTIsrael/lems/assets/51059131/68276405-6e9d-4e35-8270-65df0a4efc0a)

The call to `csrfValidator` I had in `/api/admin` was redundant. I removed it.

johnmeshulam commented:
Thanks, aproved
